WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift

```diff
@@ -0,0 +1,257 @@
+
+    private let viewModel: ReprintShippingLabelViewModel
+    private let rows: [Row]
+
+    private var selectedPaperSize: ShippingLabelPaperSize?
+
+    private var cancellables = Set<AnyCancellable>()
+
+    init(shippingLabel: ShippingLabel) {
+        self.viewModel = ReprintShippingLabelViewModel(shippingLabel: shippingLabel)
+        self.rows = [.headerText, .infoText, .paperSize, .spacer, .paperSizeOptions, .printingInstructions]
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    deinit {
+        cancellables.forEach {
+            $0.cancel()
+        }
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        configureNavigationBar()
+        configureTableView()
+        configureReprintButton()
+        observeSelectedPaperSize()
+    }
+
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+
+        tableView.updateHeaderHeight()
+        tableView.updateFooterHeight()
+    }
+}
+
+// MARK: Action Handling
+private extension ReprintShippingLabelViewController {
+    func reprintShippingLabel() {
+        // TODO-2169: reprint action
+    }
+}
+
+// MARK: Configuration
+private extension ReprintShippingLabelViewController {
+    func configureNavigationBar() {
+        navigationItem.title = Localization.navigationBarTitle
+    }
+
+    func configureTableView() {
+        tableView.dataSource = self
+        tableView.delegate = self
+        registerTableViewCells()
+
+        tableView.tableFooterView = UIView()
+        tableView.backgroundColor = .basicBackground
+    }
+
+    func registerTableViewCells() {
+        for row in Row.allCases {
+            tableView.registerNib(for: row.type)
+        }
+    }
+
+    func configureReprintButton() {
+        reprintButton.applyPrimaryButtonStyle()
+        reprintButton.setTitle(Localization.reprintButtonTitle, for: .normal)
+        reprintButton.on(.touchUpInside) { [weak self] _ in
+            self?.reprintShippingLabel()
+        }
+    }
+
+    func observeSelectedPaperSize() {
+        viewModel.loadShippingLabelSettingsForDefaultPaperSize()
+        viewModel.$selectedPaperSize.sink { [weak self] paperSize in
+            guard let self = self else { return }
+            self.selectedPaperSize = paperSize
+            self.tableView.reloadData()
+            self.reprintButton.isEnabled = paperSize != nil
+        }.store(in: &cancellables)
+    }
+}
+
+// MARK: UITableViewDataSource
+extension ReprintShippingLabelViewController: UITableViewDataSource {
+    func numberOfSections(in tableView: UITableView) -> Int {
+        1
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        rows.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let row = rows[indexPath.row]
```

Comment:
If there's one thing I hate, it's “invalid number of rows” exceptions. 😄 I think we should still consider using `safe` whenever possible. 

```suggestion
        guard let row = rows[safe: indexPath.row] else {
            return UITableViewCell()
        }
```

But up to you. It's unlikely. We can always fix it if we ever get a crash later on the road.

Reply:
Hah right, it doesn't hurt to be safer. Updated in: https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/79499744b164335f3b80cad620152502c4d1a96b